cdap-data-fabric/src/main/java/co/cask/cdap/data2/dataset2/lib/table/hbase/HBaseTable.java

```diff
@@ -317,37 +323,67 @@ private boolean hbaseFlush(List<Mutation> mutations)
     return false;
   }
 
-  private PutBuilder getPut(PutBuilder existing, byte[] row) {
+  private PutBuilder getPut(PutBuilder existing, byte[] row, @Nullable byte[] txId) {
     if (existing != null) {
       return existing;
     }
-    return tableUtil.buildPut(row)
-      .setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
+    PutBuilder putBuilder = tableUtil.buildPut(row);
+    if (txId != null) {
+      putBuilder.setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis)
+        .setAttribute(TX_ID, txId);
+    }
+     return putBuilder;
   }
 
   private PutBuilder getPutForIncrement(PutBuilder existing, byte[] row) {
     if (existing != null) {
       return existing;
     }
     return tableUtil.buildPut(row)
-      .setAttribute(DELTA_WRITE, Bytes.toBytes(true))
-      .setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
+      .setAttribute(DELTA_WRITE, Bytes.toBytes(true));
   }
 
-  private IncrementBuilder getIncrement(IncrementBuilder existing, byte[] row, boolean transactional)
+  private IncrementBuilder getIncrement(IncrementBuilder existing, byte[] row, @Nullable byte[] txId,
+                                        @Nullable byte[] txWritePointer)
     throws IOException {
     if (existing != null) {
       return existing;
     }
     IncrementBuilder builder = tableUtil.buildIncrement(row).setAttribute(DELTA_WRITE, Bytes.toBytes(true));
-    builder.setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
-    if (transactional) {
-      builder.setAttribute(WRITE_POINTER, Bytes.toBytes(tx.getWritePointer()));
+    if (txId != null) {
+      builder.setAttribute(TX_MAX_LIFETIME_MILLIS_KEY, txMaxLifetimeMillis);
+      builder.setAttribute(WRITE_POINTER, txWritePointer);
```

Comment:
are these two needed? They are already part of getEncodedTx()

Reply:
The Increment co-processor does not decode the transaction. It just passes it on to the child operations it creates.

Reply:
You'll have to change the IncrementHandler to either enforce the tx lifetime, or to pass on these attributes to the (get and?) put operations it performs.

Reply:
Good catch. Updated the code.